### PR TITLE
Gpr smooth logic

### PR DESCRIPTION
**Pull request by talonmyburgh**
@landmanbester please review. I have attached a Python script with which to test the behaviour of the GPR smoothing kernel against the Gaussian convolution process.

You'll first need to `pip install inspect-dynspec`, but then you can call it as:
`python ./test_smoothing_kernels.py --root <path-to-rims-output> --output <path-to-write-plots-to> --test-ones`

You can omit the `root` args and default to using my RIMS output for the Parrot Flare on `bbking`. `test-ones` replaces target dynspec, W and W2 with ones... but does not replace the off-targets -> so you'll see some values in the `var_e`.

Without `test-ones`, GPR does not match `Gaussian convolution`. Likely because it just smoothes the contents of `A.10` where the Gaussian Convolution smooths `A.10 / A.12` and then divides that smoothed result by a smoothed `A.12`.

Thanks.
[test_smoothing_kernels.py](https://github.com/user-attachments/files/24104572/test_smoothing_kernels.py)

### CODE REVIEW

File: inspect_dynspec/gpr_smooth.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        R_T @ y_obs: scatter residuals back into full image.
+        Returns array of shape (nx * ny,), which you can reshape to (nx, ny).
+        """
+        # start with zeros in flattened space
+        full = jnp.zeros(self.mask_flat.shape)
+        # scatter observed values back
+        full = full.at[self.mask_flat].set(y_obs)
+        return full  # still flattened
+
+    def adjoint_image(self, y_obs: jnp.ndarray) -> jnp.ndarray:
```
Comment (landmanbester) [inline inspect_dynspec/gpr_smooth.py]: Is there a good reason for adding the reshape as a separate method? If not, consider making it part of `adjoint` so that it really is the adjoint of the `forward` operator

File: inspect_dynspec/gpr_smooth.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    weights = jnp.ones_like(data) if weights is None else weights
+
+    data_flat = data.ravel()
+    prec_flat = weights.ravel()
+
+    y_obs = R.forward(data_flat)  # observed data
+    prec_obs = R.forward(prec_flat)  # observed precisions
+
+    # Σ = diag(1/prec_flat)  ⇒  σ_obs = 1/√prec_obs
+    sigma_obs = 1.0 / jnp.sqrt(jnp.where(prec_obs == 0, 1, prec_obs))
```
Comment (landmanbester) [inline inspect_dynspec/gpr_smooth.py]: Ideally the mask operator should remove entries with zero weight so the `where` shouldn't be necessary. Might be better to raise an error here if any `prec_obs` are zero

File: inspect_dynspec/gpr_smooth.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+    y_white = y_obs * prec_obs  # whitened observations
```
Comment (landmanbester) [inline inspect_dynspec/gpr_smooth.py]: Technically a whitened variable is normalised by its standard deviation (not variance) so calling `y_white` whitened is not really correct. Maybe weighted?

File: inspect_dynspec/gpr_smooth.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    Ls = (Lv, Lt)
+    Ls_T = tuple(A.T for A in Ls)
+
+    # (I + Lᵀ Rᵀ Σ⁻¹ R L) * Eta = Lᵀ Rᵀ Σ⁻¹ * data
+    # A * Eta = b:
+    # A = (I + Lᵀ Rᵀ Σ⁻¹ R L) is the operator we want to apply
+    # However now Σ⁻¹ = diag(prec_flat) here (heteroscedastic noise),
+    A_matvec = make_A_matvec(Ls, R, prec_flat)
+
+    # b = kron_mv([A.T for A in Ls[::-1]], RT_Sinv_y)
```
Comment (landmanbester) [inline inspect_dynspec/gpr_smooth.py]: Debug comment?

File: inspect_dynspec/gpr_smooth.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        # 1) x = kron_mv(Ls, z)
+        x_flat = kron_mv(Ls, z)
+        # 2) Rx = mask.forward(x_flat)
+        Rx = mask.forward(x_flat)
+        # 3) Σ⁻¹ Rx  ← use per-pixel precisions at observed locs
+        prec_obs = mask.forward(prec_flat)
+        Sinv_Rx = Rx * prec_obs
+        # 4) Rᵀ (Σ⁻¹ Rx)
+        RT_Sinv_Rx = mask.adjoint(Sinv_Rx)
+        # 5) Lᵀ term: kron_mv([A.T for A in Ls[::-1]], RT_Sinv_Rx)
```
Comment (landmanbester) [inline inspect_dynspec/gpr_smooth.py]: Remove `[::-1]` in comment

**Review by landmanbester — APPROVED** [review]
I can't see anything obviously wrong with the GPR implementation so I think it's good to go. Might be worth adding a test though. A good test would be to smooth a sample drawn from the GP prior with a representative mask applied and some noise added i.e. generate some mock data using something like the following
```
xi = np.random.randn(n_freq, n_time)  # sample from standard normal
dspec = kron_matvec(L, xi)  # apply sqrt cov matrix to get sample from GP prior
dspec_obs = R(dspec)   # apply mask
std_non_iid = np.exp(np.random.randn(n_flat))
dspec_obs += std_non_iid * np.random.randn(n_flat)
``` 
Then smooth the data with the known `L, R` and `wgt = 1/std_non_iid**2` and check if the variance computed from the samples explains the data i.e.
```
Iin = np.abs(dspec_obs - R(dspec)) <= np.sqrt(sample_var)
frac_in = np.sum(Iin) / Iin.size
assert frac_in >= 0.67  # might need to adjust lower due to noisy samples 
```
This should pass if everything works as expected since we are using the correct prior and data covariance. Ping me if uncertain

**Comment by talonmyburgh** [comment]
@landmanbester I believe it works... I ran this code:
```
import numpy as np
import jax
import jax.numpy as jnp
from jax.scipy.linalg import cholesky
# Import your specific implementation
from inspect_dynspec.gpr_smooth import gpr_smooth, rbf_kernel, kron_mv

Nv, Nt = 100, 50 
dnu_dt = (1.0, 1.0)

t_grid = jnp.linspace(0, 1, Nt)
v_grid = jnp.linspace(0, 1, Nv)
jitter = 1e-6

# Create a "smooth" sample
Kt = rbf_kernel(t_grid, lengthscale=dnu_dt[1], variance=1.0)
Kv = rbf_kernel(v_grid, lengthscale=dnu_dt[0], variance=1.0)
Lt = cholesky(Kt + jitter * jnp.eye(Nt), lower=True)
Lv = cholesky(Kv + jitter * jnp.eye(Nv), lower=True)

# Make "true" surface
key = jax.random.PRNGKey(42)
k1, k2, k3 = jax.random.split(key, 3)
xi = jax.random.normal(k1, shape=(Nv, Nt))
prior_sample = kron_mv((Lv, Lt), xi)

# iid noise
sigma = jnp.exp(0.3 * jax.random.normal(k2, shape=(Nv, Nt)))

mask_valid = jnp.ones((Nv, Nt), dtype=bool)

# Make mask - first take 4 random spots and remove 5 channels
rfi_starts = jax.random.randint(k3, shape=(4,), minval=0, maxval=Nv-5)
for start in rfi_starts:
    mask_valid = mask_valid.at[start:start+5, :].set(False)

# Then take out 2 random time chunk of 8 time steps
time_drop_start = jax.random.randint(k3, shape=(2,), minval=0, maxval=Nt-8)[0]
mask_valid = mask_valid.at[:, time_drop_start:time_drop_start+8].set(False)

print(f"Mask created. {jnp.mean(mask_valid)*100:.1f}% of data remains valid.")

noisy_surface = prior_sample + sigma * jax.random.normal(k3, shape=(Nv, Nt))

weights = 1.0 / (sigma**2)

print("Running GPR...")
smooth_surface, var_surface, _ = gpr_smooth(
    data=noisy_surface,
    mask=mask_valid,
    weights=weights,
    l_length_nu=dnu_dt[0],
    l_length_t=dnu_dt[1],
    sigma2=1.0,
    nof_weight_samples=50
)

std_deviation = jnp.sqrt(var_surface)

# How much is within 1 std deviation?
post_mask = jnp.abs(smooth_surface - prior_sample) <= std_deviation
fraction_passed = jnp.mean(post_mask)
print(f"Fraction of pixels within 1 std dev: {fraction_passed:.4f}")

assert fraction_passed >= 0.65, "Test Failed"
print("Test Passed")
```
And got:
```
Mask created. 67.2% of data remains valid.
Running GPR...
Jax CG result, info = None
Fraction of pixels within 1 std dev: 0.7826
Test Passed
```

I did "simulate" a mask because of the sizes we chose for our field `(100,50)`. Can test if extending the field more and using an actual mask makes any difference.

I assume that the fraction of pixels above `1 sigma` being significantly more that `0.67` is okay and still indicates we are testing this well?

**Comment by landmanbester** [comment]
Should be okay but it shouldn't consistently >0.67. The test looks good to me. Maybe just try with different random seeds to be sure

**Comment by talonmyburgh** [comment]
@copilot+gpt-5.3-codex resolve the merge conflicts in this pull request.

Master branch has changes we want to retain like the ability to plot/calc rm_synthesis, pol_power etc. But GPR branch has changes we want like the GPR smoothing, SNR and lightcurve plotting options and the simplified fewer plots when in debug mode vs when not.

**Comment by talonmyburgh** [comment]
Reverted that merge by copilot because it was rather terrible :)
